File: khalti-android/src/main/java/com/khalti/checkOut/Wallet/WalletModel.java
package com.khalti.checkOut.Wallet;


import java.util.HashMap;

import com.khalti.checkOut.Wallet.helper.WalletConfirmPojo;
import com.khalti.checkOut.Wallet.helper.WalletInitPojo;
import com.khalti.checkOut.api.ApiHelper;
import com.khalti.checkOut.api.Result;
import com.khalti.checkOut.helper.Config;
import com.khalti.checkOut.api.KhaltiApi;
import com.khalti.utils.EmptyUtil;
import com.khalti.utils.Store;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import kotlin.coroutines.Continuation;
import rx.Observable;
import rx.Subscriber;
import rx.subjects.PublishSubject;
import rx.subscriptions.CompositeSubscription;

public class WalletModel implements WalletContract.Model {
    private KhaltiApi khaltiService;
    private ApiHelper apiHelper;
    private WalletInitPojo walletInit;
    private CompositeSubscription compositeSubscription;

    WalletModel() {
        khaltiService = ApiHelper.apiBuilder();
        apiHelper = new ApiHelper();
        compositeSubscription = new CompositeSubscription();
    }

    WalletModel(KhaltiApi mockedKhaltiService) {
        khaltiService = mockedKhaltiService;
        compositeSubscription = new CompositeSubscription();
    }


    @Nullable
    @Override
    public Object initiatePayment(@NotNull String mobile, @NotNull Config config, @NotNull Continuation<? super Result<WalletInitPojo>> continuation) {
        return null;
    }

    @Nullable
    @Override
    public Object confirmPayment(@NotNull String confirmationCode, @NotNull String transactionPIN, @NotNull String token, @NotNull Continuation<? super Result<WalletConfirmPojo>> continuation) {
        return null;
    }
}

File: khalti-android/src/main/java/com/khalti/checkOut/wallet/WalletModel.kt
package com.khalti.checkOut.wallet

import com.khalti.checkOut.wallet.helper.WalletConfirmPojo
import com.khalti.checkOut.wallet.helper.WalletInitPojo
import com.khalti.checkOut.api.ApiHelper_
import com.khalti.checkOut.api.KhaltiApi_
import com.khalti.checkOut.api.Result
import com.khalti.checkOut.api.Urls
import com.khalti.checkOut.helper.Config
import com.khalti.utils.EmptyUtil
import com.khalti.utils.Store

class WalletModel(mockedKhaltiApi: KhaltiApi_? = null) : WalletContract.Model {

    private var khaltiApi: KhaltiApi_ = if (EmptyUtil.isNotNull(mockedKhaltiApi)) {
        mockedKhaltiApi!!
    } else {
        ApiHelper_.apiBuilder()
    }
    private var apiHelper: ApiHelper_ = ApiHelper_()

    override suspend fun initiatePayment(mobile: String, config: Config): Result<WalletInitPojo> {

        val dataMap = HashMap<String, Any>()
        dataMap["public_key"] = config.publicKey
        dataMap["product_identity"] = config.productId
        dataMap["product_name"] = config.productName
        if (EmptyUtil.isNotNull(config.productUrl) && EmptyUtil.isNotEmpty(config.productUrl)) {
            dataMap["product_url"] = config.productUrl!!
        }
        dataMap["amount"] = config.amount
        dataMap["mobile"] = mobile
        if (EmptyUtil.isNotNull(config.additionalData)) {
            dataMap.putAll(config.additionalData!!)
        }

        return apiHelper.callApi(khaltiApi.initiatePayment(Urls.WALLET_INITIATE.value, dataMap))
    }

    override suspend fun confirmPayment(confirmationCode: String, transactionPIN: String, token: String): Result<WalletConfirmPojo> {

        val dataMap = HashMap<String, Any>()
        dataMap["token"] = token
        dataMap["confirmation_code"] = confirmationCode
        dataMap["transaction_pin"] = transactionPIN
        dataMap["public_key"] = Store.getConfig().publicKey

        return apiHelper.callApi(khaltiApi.confirmPayment(Urls.WALLET_CONFIRM.value, dataMap))
    }
}